### PR TITLE
Image cropping

### PR DESCRIPTION
**Pull request by dansbecker**
This is my work for the evening.

It's supposed to do the random cropping, but it is getting the locations of the object after cropping wrong in a lot of images.

So there is a ton of debugging code in there right now... to be pulled out after 🐛 and 🐞  are gone.

### CODE REVIEW

File: code/data_setup/parse_imagenet_bb.py [hunk 1]
```diff
@@ -38,21 +39,22 @@ def parse_file(bbox_dir, filename):
             bounding_box = child.find('bndbox')
             xmin, xmax, ymin, ymax = parse_bb(bounding_box)
 
-            object_lst = [width, height, name, subcategory, xmin, 
+            object_lst = [filename, width, height, name, subcategory, xmin,
```
Comment (dansbecker) [inline code/data_setup/parse_imagenet_bb.py:42]: The other changed lines are just my editor doing automatic clean-up. 
This is the only line I meant to change.  Needed filename for subsequent steps.

I also use the convention x0, y0, x1, y1 instead of xmin, ymin, xmax, ymax.  At some point, we should settle on one convention.

Comment (sallamander) [inline code/data_setup/parse_imagenet_bb.py:42]: Nice, excited to dig in! I think I should have some time to play around with this this morning. 

I'm good with the `x0, y0, x1, y1` convention, so why don't we move forward with that. The other convention was what I think was used in the bbox files for imagenet, so I had just used that. No particular reason/preference for it.

File: code/data_setup/image_cropping.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            crop_coords, new_box_coords = get_random_crop_coords(box_coords,
+                                                                 img_coords,
+                                                                 output_img_width=output_width,
+                                                                 output_img_height=output_height)
+            print("------------")
+            print(img_coords)
+            print(box_coords)
+            print(crop_coords)
+            print(new_box_coords)
+            cropped_img = original_img[crop_coords.x0:crop_coords.x1,
```
Comment (dansbecker) [inline code/data_setup/image_cropping.py]: This was the error, and it's an error I've made many times before.

⚠️  You index into an image as `img[y0, x0, y1, x1]` instead of `img[x0, y0, x1, y1]`. Because the array is indexed as rows before columns... and a row number corresponds to a height in the image.

File: motion_tracker/data_setup/image_cropping.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    max_box_frac_of_height = 0.66
+    images_successfully_downloaded = set(os.listdir('data/imagenet/images'))
+    bbox_df = (pd.read_csv('work/imagenet/parsed_bb.csv')
+                    .assign(box_height = lambda df: df.ymax - df.ymin,
+                            box_width = lambda df: df.xmax - df.xmin)
+                    .assign(box_frac_of_height = lambda df: df.box_height / df.height,
+                            box_frac_of_width = lambda df:  df.box_width / df.width)
+                    .query('filename in @images_successfully_downloaded')
+                    .query('box_frac_of_height < @max_box_frac_of_height')
+                    .query('box_frac_of_width < @max_box_frac_of_width'))
```
Comment (sallamander) [inline motion_tracker/data_setup/image_cropping.py]: Should this be a `>=` in the last two queries? They said that the bounding box should fill **at least** 66% of the...

File: motion_tracker/data_setup/image_cropping.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    cropped_area_coords = Coords(cropped_area_x0, cropped_area_y0,
+                                 cropped_area_x1, cropped_area_y1)
+
+    box_after_crop = update_box_after_crop(box_coords, cropped_area_coords)
+    return cropped_area_coords, box_after_crop
+
+def read_bbox_data(raw_image_dir, parsed_bb_path='work/imagenet/parsed_bb.csv'):
+
+    # paper says they use .66, but they also have search area that is twice bounding
+    # box. So, 0.5 may be more appropriate.
```
Comment (sallamander) [inline motion_tracker/data_setup/image_cropping.py]: Yeah, we'll have to play around with this.

**Comment by sallamander** [comment]
Good stuff! Feel free to merge. I didn't see anything jump out at me.

**Comment by dansbecker** [comment]
@sallamander There's another commit that includes minor refactoring and the addition of a generator for crops from the imagenet images.  You might want to look at it.

There are a bunch of images that I can't open, either in opencv or anything else.  I'm probably going to write a script that either moves those to a new place... or creates a list of them and just deletes them.

But life will be a little simpler if we get those out of the directory we read images from.

Also, creating a batch of 50 training images currently takes 1 second.  I think that's problematically slow.  It will take more than 30 minutes per epoch just to serve the data to the model.  I'm hoping to look tonight at what can be done to speed that up.

Anyway, this is worth a second look.

**Comment by sallamander** [comment]
👍 For sure, I'll look it over closely either tonight or tomorrow morning. I'll especially keep an eye out for things we can do to speed anything up. 

Out of curiosity, can you give me an image # you can't open? Do they open with your Finder on mac, or anything like that? Is it that you can't open them programmatically?

In general, when I looked it over earlier I thought that some refactoring might be a good idea, but also was on the fence on whether it was me just being picky or maybe picking things out just because it wasn't the way I would write it. In the future, I'll go ahead and comment on these things, and you can feel free to ignore them if that's the case.

**Comment by dansbecker** [comment]
Cool.  I did some refactoring, and I'm sure there's more that can be done.

**Comment by sallamander** [comment]
Moving on to refactor on my end and add the alov image generator. 💂

File: motion_tracker/utils/image_cropping.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    bbox_df = (pd.read_csv('work/imagenet/parsed_bb.csv')
+                    .assign(box_height = lambda df: df.ymax - df.ymin,
+                            box_width = lambda df: df.xmax - df.xmin)
+                    .assign(box_frac_of_height = lambda df: df.box_height / df.height,
+                            box_frac_of_width = lambda df:  df.box_width / df.width)
+                    .query('filename in @images_successfully_downloaded')
+                    .query('box_frac_of_height < @max_box_frac_of_height')
+                    .query('box_frac_of_width < @max_box_frac_of_width'))
+    return bbox_df
+
```
Comment (sallamander) [inline motion_tracker/utils/image_cropping.py:192]: What if we did this outside of the generator? At some point when we have all the images downloaded that we need to, this becomes a 1 time operation, yeah? This will shave a little bit of time off the generator setup.

Comment (dansbecker) [inline motion_tracker/utils/image_cropping.py:192]: Your thinking we would do this, save the output, and then load that saved
file when we start the generator?

That seems like a good idea.

File: motion_tracker/utils/image_cropping.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    output_width = 256
+    output_height = 256
+
+    start_img_data = []
+    start_box_data = []
+    next_img_data = []
+    next_box_data = []
+
+    while True:
+        batch_df = img_metadata.sample(batch_size)
```
Comment (sallamander) [inline motion_tracker/utils/image_cropping.py:233]: Will we still do a `sample` like this when we are training? I imagined effectively iterating over all of the training images after randomly shuffling them, or something like that. I getting rid of this `sample` might speed things up as well. The code under the hood of that is somewhat involved, and I could imagine that iterating over the training images will add some small gains.

File: motion_tracker/utils/image_cropping.py (hunk 5 above)
Comment (sallamander) [inline motion_tracker/utils/image_cropping.py:192]: Yep, exactly.

Comment (dansbecker) [inline motion_tracker/utils/image_cropping.py:192]: Cool. Want to add that to the refactoring PR.

Comment (sallamander) [inline motion_tracker/utils/image_cropping.py:192]: Definitely. I'm playing around with stuff currently and trying to figure out where we can get gains.

File: motion_tracker/utils/image_cropping.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    '''Utility class to hold coords and automatically calculate height, width and center
+
+    x0, y0, x1 and y1 are all integers
+    '''
+
+    def __init__(self, x0, y0, x1, y1):
+        self.x0 = int(x0) # convert to int here. Otherwise shows up in many places
+        self.y0 = int(y0)
+        self.x1 = int(x1)
+        self.y1 = int(y1)
```
Comment (sallamander) [inline motion_tracker/utils/image_cropping.py]: Do I understand correctly that a lack of casting to `int` here really just causes problems for drawing bounding boxes during visualization? When I remove this and play with stuff, it doesn't cause any issues until I try to use your `show_stages_of_random_crop` function. So, I'm wondering if during training this is necessary or will be an issue? Removing it would save four calls to `int` here, and maybe more importantly (although I still don't see this having a big effect) is that we are throwing away some information by casting our bounding boxes to ints where previously they were floats. This change in the end probably won't matter, but it's maybe something to think about.

File: motion_tracker/utils/image_cropping.py (hunk 6 above)
Comment (dansbecker) [inline motion_tracker/utils/image_cropping.py:233]: I don't know. I was just about to make this sample over images, then do 10 crops for that image.

That should give a big speed-up, since more than 90% of the time in the generator is just loading the images.  Whether we iterate sequentially over images or sample over them doesn't seem important to me.

The [generators in keras](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py) and in tensorflow.learn have an argument for whether you want to shuffle or take the data in order. 

But I don't expect it to make a big difference either way, and I was just going to choose based on what created the simplest code... which I think is to randomly select images.

Comment (sallamander) [inline motion_tracker/utils/image_cropping.py:233]: 👍
